### ISSUE TITLE
Overlays cause screen to flash (modeset?) when in a full screen game (regression)

### ISSUE THREAD

**Issue by marcelohdez** [swaywm/sway#8440]
### Please fill out the following:
- **Sway Version:** 1.10
- **Debug Log:** https://gist.github.com/marcelohdez/513599d7f412b87b2f51c65584e012da
- **Configuration File:** Works with default config
- **Stack Trace:**
- **Description:**
  - When in a full screen game (not in e.g. Firefox for some reason), any overlay (from Mako notifications, to Steam notifications) will cause the screen to flash black when it is shown, and again when it goes away
  - Tested with the game Ancestors: The Humankind Odyssey
  - This does not occur in Sway 1.9
  - Video from phone (not visible in screen recording, although screen recording has another issue I will create after this one):

https://github.com/user-attachments/assets/375f828d-6899-4388-a76a-f33cb375e115

**Comment by marcelohdez** [comment]
After some more testing it seems to occur _only_ in Steam/Proton games?
- Firefox (wayland) does not have this issue
- Minecraft (xwayland) through Prism Launcher does not either
- Far Cry Primal on Bottles (wine, xwayland) does not either

**however**, Skyrim through Steam Proton _does_ have the issue.

Will try to bisect soon

**Comment by emersion** [comment]
Does `WLR_SCENE_DISABLE_DIRECT_SCANOUT=1` help?

**Comment by marcelohdez** [comment]
Setting that variable removes the black screen, however it still stutters when the overlay is shown

After bisecting it seems the issue begins on commit 19ca790a9f304bb138e531719a8a42f145f835f9

**Comment by kennylevinsen** [comment]
Can you test if https://github.com/swaywm/sway/pull/8447 resolves the issue?

The commit you found is a fix for a regression introduced in https://github.com/swaywm/sway/commit/1e0031781fc9283db7096aba34deca5503c2ab91, but a minor behavior difference was still left behind.

**Comment by marcelohdez** [comment]
I cherry-picked those commits on top of `v1.10` instead of building `master` as Fedora does not have wlroots 0.19 however they _did_ make the issue much better. With those commits appplied the black screens are gone in-game and so is some stutters that would affect the overlays such as Steam's FPS counter and their in-game overlay.

Although the issue still persists during the loading screen of Ancestors for some reason, where it did not occur in Sway 1.9 (tried again just now), haven't tested other games yet.

**Comment by kennylevinsen** [comment]
For the other glitch sources, is https://github.com/swaywm/sway/commit/19ca790a9f304bb138e531719a8a42f145f835f9 still the first commit showing the issue? If so, does it happen *before* the original regression (https://github.com/swaywm/sway/commit/1e0031781fc9283db7096aba34deca5503c2ab91)?

**Comment by marcelohdez** [comment]
I have not tested those extensively, but I have found a way to reproduce the issue for free so someone can test locally:

- Install Bottles
- Create a new Gaming bottle (`+` on top left) and go into it
- Download the windows zip of this game: https://roarkdev.itch.io/raot
- Unzip it anywhere Bottles can see (if using Flatpak and outside the bottle's directory, Flatseal can give Bottles permission to see it)
- "Run Executable..." and select the `raot.exe` file
- When inside the game (should be fullscreen), you can ignore firewall warning etc. and host a game, select any team and character
- `esc` to go to settings, turn graphics to high, anti-aliasing, bloom etc, put system in power-saver, anything to make frame times inconsistent (leave fps limit != your screen's hz)
- You should see stutters and black screens as long as VSync is off (maybe a clue?) and running around `WASD`/dashing `Q` making an overlay popup every now and then, e.g.: 
```bash
while true; do notify-send hi -t 2000; sleep 3; done
```

**Comment by ErikReider** [comment]
Is this related or a separate issue? (Video in SwayFX, but also affects Sway 1.10)

https://github.com/user-attachments/assets/288994a1-ab4f-4989-a65e-dc3716f3b585

A lot of Steam games like CS2 and Lethal Company stutter when not using direct scanout. So fullscreening the games fixes this, but receiving a notification in the overlay layer and trying to screen share disables direct scanout.

Weirdly enough, some games don't experience this like Half Life and CSGO (both Source 1) which also runs natively XWayland like CS2 (Source 2)

**Comment by kennylevinsen** [comment]
Can you try:
1. `WLR_RENDERER=vulkan` if you haven't already
2. `WLR_RENDER_NO_EXPLICIT_SYNC=1` on both vulkan and gles2 renderer
3. `WLR_SCENE_DEBUG_DAMAGE=rerender`, in case above doesn't work and it ends up being a damage issue

**Comment by ErikReider** [comment]
> Can you try:
> 
>     1. `WLR_RENDERER=vulkan` if you haven't already
> 
>     2. `WLR_RENDER_NO_EXPLICIT_SYNC=1` on both vulkan and gles2 renderer
> 
>     3. `WLR_SCENE_DEBUG_DAMAGE=rerender`, in case above doesn't work and it ends up being a damage issue

Using the vulkan renderer fixes the issues! I opened an issue [here](https://gitlab.freedesktop.org/wlroots/wlroots/-/issues/3950)

**Comment by pipsuko** [comment]
I am having a similar issue with the notification overlay and screenshare when a game is in fullscreen. 

When a game is in fullscreen and an overlay is displayed such as a notification the screen will flicker black and will stutter for the duration the overlay is present. Screensharing also causes the issue but on a worse scale, ***even without an overlay present***. Having MangoHud enabled while the issue is occurring demonstrates the extremely strange behavior (video attached). It is primarily present in games running through wine, but it sometimes affects native Linux games. 

I have tried every environment variable [mentioned above](https://github.com/swaywm/sway/issues/8440#issuecomment-2690760513), they did not resolve the issue at all. I have also tried using sway-master, that did not work. What did work was downgrading Sway to version 1.9 or using gamescope.

https://github.com/user-attachments/assets/991d3566-943c-4657-812f-300827f0391c

**Comment by marcelohdez** [comment]
Where did you set the environment variable? Would like to add that for me personally, using the Vulkan renderer fixes the issue on 1.10.1

**Comment by pipsuko** [comment]
> Where did you set the environment variable? Would like to add that for me personally, using the Vulkan renderer fixes the issue on 1.10.1

I set them in the TTY before starting sway. What is your experience with screensharing?

**Comment by marcelohdez** [comment]
> I set them in the TTY before starting sway. What is your experience with screensharing?

More testing is needed however from simply playing the game I did initially in issue description (Ancestors: A Humankind Odyssey) and screen sharing entire screen using this website https://www.webrtc-experiment.com/Pluginfree-Screen-Sharing on Vulkan renderer everything seems fine, while reverting back to regular renderer the frame-repeating and overall weirdness comes back.

Could you check inside your sway session if `echo $WLR_RENDERER` returns `vulkan`? Since I use GDM on my Fedora install I just edited the `/usr/share/wayland-sessions/sway.desktop` file by hand and added`Exec=env WLR_RENDERER=vulkan sway`

**Comment by kennylevinsen** [comment]
The flickering looks like sync or damage issues.

If you try master again, check the sway debug output to see what renderer is used - attach it here if you want us to look at it.

> Screensharing also causes the issue but on a worse scale

That is a different issue, see https://gitlab.freedesktop.org/wlroots/wlroots/-/issues/3953 - the problem is that every time the screen share tool asks for a frame, direct scanout is disabled, then almost immediately afterwards we render a frame where the screenshare tool *isn't* asking for a frame so we enable it again.

It is worked around by `WLR_SCENE_DISABLE_DIRECT_SCANOUT=1`, or this hack MR: https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/5013

**Comment by ErikReider** [comment]
> > I set them in the TTY before starting sway. What is your experience with screensharing?
> 
> More testing is needed however from simply playing the game I did initially in issue description (Ancestors: A Humankind Odyssey) and screen sharing entire screen using this website https://www.webrtc-experiment.com/Pluginfree-Screen-Sharing on Vulkan renderer everything seems fine, while reverting back to regular renderer the frame-repeating and overall weirdness comes back.
> 
> Could you check inside your sway session if `echo $WLR_RENDERER` returns `vulkan`? Since I use GDM on my Fedora install I just edited the `/usr/share/wayland-sessions/sway.desktop` file by hand and added`Exec=env WLR_RENDERER=vulkan sway`

For me, enabling drm sync fixes this, even with the gl renderer in sway master

**Comment by pipsuko** [comment]
> Could you check inside your sway session if `echo $WLR_RENDERER` returns `vulkan`?

It does.

> If you try master again, check the sway debug output to see what renderer is used - attach it here if you want us to look at it.

I made two debug files. One for the screen-share and one just receiving notifications, both with `WLR_RENDERER=vulkan` `WLR_RENDER_NO_EXPLICIT_SYNC=1` and `WLR_SCENE_DEBUG_DAMAGE=rerender` and on sway-master.

**Screenshare debug:**

[sway.log](https://github.com/user-attachments/files/19493060/sway.log)

Steps taken:
1. Opened terminal to double check environment variables were set properly
2. Opened discord app to turn on screenshare
3. Started steam and launched game to test (FFXIV, but occurs with any wine game)
4. Let it run for about a minute, and then closed Sway

**Notification debug:**

[sway2.log](https://github.com/user-attachments/files/19493097/sway2.log)

Steps taken:
1.  Launched discord app to receive notifications
2.  Started steam and launched game to test
3. Received 6 notifications
4. Closed Sway

In my situation the vulkan renderer does not resolve the screen share issue. Neither does it resolve the stuttering that occurs when a notification overlay is active. However, it does prevent the black screen flicker that occurs when a notification overlay first appears.

**Comment by kennylevinsen** [comment]
> In my situation the vulkan renderer does not resolve the screen share issue.

That is as expected - as mentioned, this is a different issue worked around by setting `WLR_SCENE_DISABLE_DIRECT_SCANOUT=1` (or by applying one of the temporary workaround MR's). You didn't set that variable.

> Neither does it resolve the stuttering that occurs when a notification overlay is active. However, it does prevent the black screen flicker that occurs when a notification overlay first appears.

In your log:
1. Every few seconds, a notification is created which causes direct scanout to be disabled. This means that we send dmabuf feedback without scanout recommendations.
2. The game might be reallocating its buffers as a result of these recommendations even though the current buffers compatible with scanout are also just fine for composition.
3. A few seconds later the notifcation goes away, allowing an attempt at direct scanout. This causes us to send out dmabuf feedback *with* scanout recommendations (which differ from those without).
4. Attempts to scan out fail due to the buffer format modifier not being compatible with the hardware requirements for scanout - likely because they followed the recommendations from point 1.
5. A few frames later, direct scanout is successful - likely because the client (your game) allocated new buffers to fit our dmabuf feedback recommendations
6. Repeat for next notification.

If the game re-allocates buffers every time the notification pops up, that would likely cause stutters.

Disabling direct scanout (the flag you didn't set) should also fix that for now.
